sycl/include/CL/sycl/usm/usm_allocator.hpp

```diff
@@ -148,6 +149,20 @@ class usm_allocator {
     }
   }
 
+  template <class U, usm::alloc AllocKindU, size_t AlignmentU>
+  friend bool operator==(const usm_allocator<T, AllocKind, Alignment> &One,
+                         const usm_allocator<U, AllocKindU, AlignmentU> &Two) {
+    return ((AllocKind == AllocKindU) && (One.MContext == Two.MContext) &&
+            (One.MDevice == Two.MDevice));
+  }
+
+  template <class U, usm::alloc AllocKindU, size_t AlignmentU>
+  friend bool operator!=(const usm_allocator<T, AllocKind, Alignment> &One,
+                         const usm_allocator<U, AllocKindU, AlignmentU> &Two) {
+    return !((AllocKind == AllocKindU) && (One.MContext == Two.MContext) &&
+             (One.MDevice == Two.MDevice));
+  }
+
 private:
   constexpr size_t getAlignment() const {
     /*
```

Comment:
0 is treated as "default - do something legal"

Reply:
Yes the template argument is fine as-is. What's not OK is that you only pass `getAlignment`  to `aligned_alloc` and it might be zero. `Alignment` is 0 and `T` is over-aligned there is no way for `aligned_alloc` to know.

You can either
- Change the template argument
- Uncomment the next few lines
- Or pass T to `aligned_alloc`